File: backend/laravel/load_test.php
#!/usr/bin/env php
<?php
/**
 * Нагрузочный тест для проверки масштабируемости системы.
 * Тестирует систему с ~100 зонами, проверяет latency p99 и переполнение очереди.
 */

require __DIR__.'/vendor/autoload.php';

$app = require_once __DIR__.'/bootstrap/app.php';
$app->make(\Illuminate\Contracts\Console\Kernel::class)->bootstrap();

use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\DB;
use App\Models\User;

// Конфигурация
$BASE_URL = env('APP_URL', 'http://localhost:8080');
$CONCURRENT_REQUESTS = 20; // Уменьшено для избежания rate limiting
$TOTAL_REQUESTS = 500; // Уменьшено для более реалистичного теста
$TEST_DURATION_SEC = 60;

$requestTimes = [];
$errors = [];

echo "=" . str_repeat("=", 60) . "\n";
echo "НАГРУЗОЧНОЕ ТЕСТИРОВАНИЕ СИСТЕМЫ\n";
echo "=" . str_repeat("=", 60) . "\n";
echo "Base URL: {$BASE_URL}\n";
echo "Concurrent requests: {$CONCURRENT_REQUESTS}\n";
echo "Total requests: {$TOTAL_REQUESTS}\n";
echo "Test duration: {$TEST_DURATION_SEC}s\n";
echo "=" . str_repeat("=", 60) . "\n";

// Авторизация
echo "\n[1/4] Авторизация...\n";
$user = User::where('email', 'admin@example.com')->first();
if (!$user) {
    echo "ОШИБКА: Пользователь admin@example.com не найден\n";
    exit(1);
}

// Удаляем старые токены для этого теста
$user->tokens()->where('name', 'load-test')->delete();

$token = $user->createToken('load-test')->plainTextToken;
echo "✓ Авторизация успешна\n";
echo "  Token preview: " . substr($token, 0, 20) . "...\n";

// Проверка начального состояния
echo "\n[2/4] Проверка начального состояния...\n";
$zonesCount = DB::table('zones')->count();
$nodesCount = DB::table('nodes')->count();
echo "Зон в системе: {$zonesCount}\n";
echo "Нод в системе: {$nodesCount}\n";

// Проверка метрик очереди
$queueSize = 0;
try {
    $redis = \Illuminate\Support\Facades\Redis::connection();
    $queueSize = $redis->llen('telemetry_queue') ?? 0;
    echo "Размер очереди телеметрии: {$queueSize}\n";
} catch (\Exception $e) {
    echo "Не удалось получить размер очереди: " . $e->getMessage() . "\n";
}

// Нагрузочное тестирование
echo "\n[3/4] Нагрузочное тестирование ({$TOTAL_REQUESTS} запросов)...\n";
$startTime = microtime(true);

// Используем публичные или менее защищенные endpoints для теста
$endpoints = [
    '/api/system/health', // Публичный endpoint
    '/api/system/config/full', // Требует Python service token или Sanctum
];

$requestsPerEndpoint = (int)($TOTAL_REQUESTS / count($endpoints));

// Функция для выполнения запроса
$makeRequest = function($url) use ($BASE_URL, $token, &$requestTimes, &$errors) {
    $start = microtime(true);
    try {
        $response = Http::timeout(30)
            ->withHeaders([
                'Authorization' => 'Bearer ' . $token,
                'Accept' => 'application/json'
            ])
            ->get($BASE_URL . $url);
        
        $elapsed = microtime(true) - $start;
        $requestTimes[] = $elapsed;
        
        if (!$response->successful()) {
            $errors[] = [
                'url' => $url,
                'status' => $response->status(),
                'error' => $response->body()
            ];
        }
        
        return [
            'success' => $response->successful(),
            'elapsed' => $elapsed,
            'status' => $response->status()
        ];
    } catch (\Exception $e) {
        $elapsed = microtime(true) - $start;
        $requestTimes[] = $elapsed;
        $errors[] = [
            'url' => $url,
            'error' => $e->getMessage()
        ];
        return [
            'success' => false,
            'elapsed' => $elapsed,
            'status' => 0
        ];
    }
};

// Выполняем запросы
$allRequests = [];
foreach ($endpoints as $endpoint) {
    for ($i = 0; $i < $requestsPerEndpoint; $i++) {
        $allRequests[] = $endpoint;
    }
}

// Перемешиваем для более реалистичной нагрузки
shuffle($allRequests);

// Выполняем запросы параллельно (в батчах)
$batchSize = $CONCURRENT_REQUESTS;
$batches = array_chunk($allRequests, $batchSize);

foreach ($batches as $batchIndex => $batch) {
    $results = [];
    foreach ($batch as $url) {
        $results[] = $makeRequest($url);
    }
    
    if (($batchIndex + 1) % 10 == 0) {
        $progress = min(100, (($batchIndex + 1) * $batchSize / $TOTAL_REQUESTS) * 100);
        echo "Прогресс: " . number_format($progress, 1) . "%\n";
    }
    
    // Небольшая задержка между батчами
    usleep(10000); // 10ms
}

$elapsed = microtime(true) - $startTime;

// Проверка метрик после теста
echo "\n[4/4] Проверка метрик после теста...\n";
$finalQueueSize = 0;
try {
    $redis = \Illuminate\Support\Facades\Redis::connection();
    $finalQueueSize = $redis->llen('telemetry_queue') ?? 0;
    echo "Размер очереди телеметрии после теста: {$finalQueueSize}\n";
} catch (\Exception $e) {
    echo "Не удалось получить размер очереди: " . $e->getMessage() . "\n";
}

// Результаты
echo "\n" . str_repeat("=", 60) . "\n";
echo "РЕЗУЛЬТАТЫ ТЕСТИРОВАНИЯ\n";
echo str_repeat("=", 60) . "\n";

if (!empty($requestTimes)) {
    sort($requestTimes);
    $count = count($requestTimes);
    $mean = array_sum($requestTimes) / $count;
    $median = $requestTimes[(int)($count / 2)];
    $p95 = $requestTimes[(int)($count * 0.95)];
    $p99 = $requestTimes[(int)($count * 0.99)];
    $min = min($requestTimes);
    $max = max($requestTimes);
    
    echo "\nВремя выполнения запросов:\n";
    echo "  Всего запросов: {$count}\n";
    echo "  Успешных: " . ($count - count($errors)) . "\n";
    echo "  Среднее время: " . number_format($mean, 3) . "s\n";
    echo "  Медиана (p50): " . number_format($median, 3) . "s\n";
    echo "  p95: " . number_format($p95, 3) . "s\n";
    echo "  p99: " . number_format($p99, 3) . "s\n";
    echo "  Максимум: " . number_format($max, 3) . "s\n";
    echo "  Минимум: " . number_format($min, 3) . "s\n";
    
    // Проверка целевых метрик
    echo "\n✓ p99 latency: " . number_format($p99, 3) . "s (" . number_format($p99 * 1000, 0) . "ms)\n";
    if ($p99 <= 0.5) {
        echo "  ✓ ЦЕЛЬ ДОСТИГНУТА: p99 ≤ 500ms\n";
    } else {
        echo "  ✗ ЦЕЛЬ НЕ ДОСТИГНУТА: p99 > 500ms\n";
    }
}

if (!empty($errors)) {
    echo "\nОшибки:\n";
    echo "  Всего ошибок: " . count($errors) . "\n";
    $errorTypes = [];
    foreach ($errors as $error) {
        $type = $error['error'] ?? 'unknown';
        $errorTypes[$type] = ($errorTypes[$type] ?? 0) + 1;
    }
    foreach ($errorTypes as $type => $count) {
        echo "  " . substr($type, 0, 50) . ": {$count}\n";
    }
}

if ($finalQueueSize > 1000) {
    echo "\n⚠ ПЕРЕПОЛНЕНИЕ ОЧЕРЕДИ ОБНАРУЖЕНО: {$finalQueueSize} элементов\n";
} else {
    echo "\n✓ Размер очереди в норме: {$finalQueueSize} элементов\n";
}

echo "\nОбщее время теста: " . number_format($elapsed, 2) . "s\n";
echo "RPS (запросов в секунду): " . number_format($count / $elapsed, 2) . "\n";
echo str_repeat("=", 60) . "\n";

File: backend/load_test.py
#!/usr/bin/env python3
"""
Нагрузочный тест для проверки масштабируемости системы.
Тестирует систему с ~100 зонами, проверяет latency p99 и переполнение очереди.
"""

import asyncio
import aiohttp
import time
import statistics
from typing import List, Dict, Any
import json
from datetime import datetime

# Конфигурация теста
BASE_URL = "http://localhost:8080"
API_TOKEN = None  # Будет получен через логин
CONCURRENT_REQUESTS = 50
TOTAL_REQUESTS = 1000
TEST_DURATION_SEC = 60

# Метрики
request_times: List[float] = []
errors: List[Dict[str, Any]] = []


async def login(session: aiohttp.ClientSession) -> str:
    """Авторизация и получение токена."""
    login_data = {
        "email": "admin@example.com",
        "password": "password"  # Изменить на реальный пароль
    }
    
    try:
        async with session.post(f"{BASE_URL}/api/auth/login", json=login_data) as resp:
            if resp.status == 200:
                data = await resp.json()
                return data.get("token") or data.get("access_token")
    except Exception as e:
        print(f"Ошибка авторизации: {e}")
    
    return None


async def make_request(session: aiohttp.ClientSession, url: str, token: str = None) -> Dict[str, Any]:
    """Выполнить HTTP запрос и измерить время."""
    headers = {}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    
    start_time = time.time()
    try:
        async with session.get(url, headers=headers, timeout=aiohttp.ClientTimeout(total=30)) as resp:
            elapsed = time.time() - start_time
            body = await resp.text()
            
            return {
                "status": resp.status,
                "elapsed": elapsed,
                "success": 200 <= resp.status < 300,
                "error": None
            }
    except asyncio.TimeoutError:
        elapsed = time.time() - start_time
        return {
            "status": 0,
            "elapsed": elapsed,
            "success": False,
            "error": "timeout"
        }
    except Exception as e:
        elapsed = time.time() - start_time
        return {
            "status": 0,
            "elapsed": elapsed,
            "success": False,
            "error": str(e)
        }


async def test_endpoint(session: aiohttp.ClientSession, endpoint: str, token: str, num_requests: int):
    """Тестировать конкретный endpoint."""
    url = f"{BASE_URL}{endpoint}"
    tasks = []
    
    for _ in range(num_requests):
        task = make_request(session, url, token)
        tasks.append(task)
    
    results = await asyncio.gather(*tasks, return_exceptions=True)
    
    for result in results:
        if isinstance(result, Exception):
            errors.append({"endpoint": endpoint, "error": str(result)})
        elif isinstance(result, dict):
            request_times.append(result["elapsed"])
            if not result["success"]:
                errors.append({
                    "endpoint": endpoint,
                    "status": result["status"],
                    "error": result.get("error")
                })


async def check_queue_metrics(session: aiohttp.ClientSession) -> Dict[str, Any]:
    """Проверить метрики очереди из Prometheus."""
    try:
        # Попробуем получить метрики из Prometheus
        prometheus_url = "http://localhost:9090/api/v1/query"
        
        queries = [
            "telemetry_queue_size",
            "telemetry_queue_overflow_total",
            "telemetry_dropped_total",
            "zone_processing_time_seconds",
            "zone_processing_errors_total"
        ]
        
        metrics = {}
        for query in queries:
            try:
                async with session.get(prometheus_url, params={"query": query}) as resp:
                    if resp.status == 200:
                        data = await resp.json()
                        if data.get("status") == "success" and data.get("data", {}).get("result"):
                            metrics[query] = data["data"]["result"]
            except:
                pass
        
        return metrics
    except Exception as e:
        print(f"Ошибка получения метрик: {e}")
        return {}


def calculate_percentile(data: List[float], percentile: float) -> float:
    """Вычислить перцентиль."""
    if not data:
        return 0.0
    sorted_data = sorted(data)
    index = int(len(sorted_data) * percentile / 100)
    return sorted_data[min(index, len(sorted_data) - 1)]


async def main():
    """Основная функция нагрузочного теста."""
    print("=" * 60)
    print("НАГРУЗОЧНОЕ ТЕСТИРОВАНИЕ СИСТЕМЫ")
    print("=" * 60)
    print(f"Base URL: {BASE_URL}")
    print(f"Concurrent requests: {CONCURRENT_REQUESTS}")
    print(f"Total requests: {TOTAL_REQUESTS}")
    print(f"Test duration: {TEST_DURATION_SEC}s")
    print("=" * 60)
    
    async with aiohttp.ClientSession() as session:
        # Авторизация
        print("\n[1/4] Авторизация...")
        token = await login(session)
        if not token:
            print("ОШИБКА: Не удалось авторизоваться")
            return
        print("✓ Авторизация успешна")
        
        # Проверка начального состояния
        print("\n[2/4] Проверка начального состояния...")
        initial_metrics = await check_queue_metrics(session)
        print(f"Начальные метрики очереди: {len(initial_metrics)} метрик")
        
        # Нагрузочное тестирование
        print(f"\n[3/4] Нагрузочное тестирование ({TOTAL_REQUESTS} запросов)...")
        start_time = time.time()
        
        # Тестируем разные endpoints
        endpoints = [
            "/api/zones",
            "/api/nodes",
            "/api/system/config/full",
        ]
        
        requests_per_endpoint = TOTAL_REQUESTS // len(endpoints)
        
        tasks = []
        for endpoint in endpoints:
            task = test_endpoint(session, endpoint, token, requests_per_endpoint)
            tasks.append(task)
        
        await asyncio.gather(*tasks)
        
        elapsed = time.time() - start_time
        
        # Проверка метрик после теста
        print("\n[4/4] Проверка метрик после теста...")
        final_metrics = await check_queue_metrics(session)
        
        # Результаты
        print("\n" + "=" * 60)
        print("РЕЗУЛЬТАТЫ ТЕСТИРОВАНИЯ")
        print("=" * 60)
        
        if request_times:
            print(f"\nВремя выполнения запросов:")
            print(f"  Всего запросов: {len(request_times)}")
            print(f"  Успешных: {sum(1 for r in request_times)}")
            print(f"  Среднее время: {statistics.mean(request_times):.3f}s")
            print(f"  Медиана (p50): {calculate_percentile(request_times, 50):.3f}s")
            print(f"  p95: {calculate_percentile(request_times, 95):.3f}s")
            print(f"  p99: {calculate_percentile(request_times, 99):.3f}s")
            print(f"  Максимум: {max(request_times):.3f}s")
            print(f"  Минимум: {min(request_times):.3f}s")
            
            # Проверка целевых метрик
            p99 = calculate_percentile(request_times, 99)
            print(f"\n✓ p99 latency: {p99:.3f}s")
            if p99 <= 0.5:
                print("  ✓ ЦЕЛЬ ДОСТИГНУТА: p99 ≤ 500ms")
            else:
                print(f"  ✗ ЦЕЛЬ НЕ ДОСТИГНУТА: p99 > 500ms ({p99*1000:.0f}ms)")
        
        if errors:
            print(f"\nОшибки:")
            print(f"  Всего ошибок: {len(errors)}")
            error_types = {}
            for error in errors:
                error_type = error.get("error", "unknown")
                error_types[error_type] = error_types.get(error_type, 0) + 1
            for error_type, count in error_types.items():
                print(f"  {error_type}: {count}")
        
        if final_metrics:
            print(f"\nМетрики очереди:")
            for metric_name, metric_data in final_metrics.items():
                if metric_data:
                    print(f"  {metric_name}: {len(metric_data)} значений")
                    # Проверяем переполнение очереди
                    if "overflow" in metric_name or "dropped" in metric_name:
                        for result in metric_data:
                            value = float(result.get("value", [0, "0"])[1])
                            if value > 0:
                                print(f"    ⚠ ПЕРЕПОЛНЕНИЕ ОБНАРУЖЕНО: {value}")
        
        print(f"\nОбщее время теста: {elapsed:.2f}s")
        print(f"RPS (запросов в секунду): {len(request_times) / elapsed:.2f}")
        print("=" * 60)


if __name__ == "__main__":
    asyncio.run(main())

